### PR TITLE
feat: Metrics page updates

### LINKED ISSUE #1485: Deep links: update MetricsPage and MetricsPageFactory
**Description**
Update the MetricsPage and MetricsPageFactory according to the design

**Acceptance Criteria**
- [ ] MetricsPage stores the RouteName and PageParametersModel
- [ ] Update MetricsPageFactory.create() method
- [ ] Test new functionality

**Additional context**
[Deep links design](https://github.com/Flank/flank-dashboard/blob/master/metrics/web/docs/features/deep_links/01_deep_linking.md#pageparametersmodel)

### PR DESCRIPTION
**Pull request by solid-maxim**
Fixes #1485 

## Checklist

- [x] Documented
- [x] Unit tested

### CODE REVIEW

File: metrics/web/lib/common/presentation/navigation/metrics_page/metrics_page.dart [hunk 1]
```diff
@@ -21,15 +25,17 @@ class MetricsPage<T> extends Page<T> {
   /// The [maintainState] defaults to `true`.
   /// The [fullscreenDialog] defaults to `false`.
   ///
-  /// The [child], [fullscreenDialog], and [maintainState] must not be null.
+  /// The [child], [routeName], [fullscreenDialog], and [maintainState] must not be null.
   const MetricsPage({
     @required this.child,
+    @required this.routeName,
     this.maintainState = true,
     this.fullscreenDialog = false,
```
Comment (solid-vovabeloded) [inline metrics/web/lib/common/presentation/navigation/metrics_page/metrics_page.dart:33]: 🔨 
Would it be great to add a `required` annotation for these parameters since they are required too?

File: metrics/web/lib/common/presentation/navigation/metrics_page/metrics_page.dart [hunk 2]
```diff
@@ -21,15 +25,17 @@ class MetricsPage<T> extends Page<T> {
...
     LocalKey key,
     String name,
```
Comment (solid-vovabeloded) [inline metrics/web/lib/common/presentation/navigation/metrics_page/metrics_page.dart:35]: 🔨 
What do you think about avoiding passing this value to the constructor and overriding the `name` field instead?

File: metrics/web/lib/common/presentation/navigation/state/navigation_notifier.dart [hunk 3]
```diff
@@ -215,7 +215,10 @@ class NavigationNotifier extends ChangeNotifier {
 
     _currentConfiguration = newConfiguration;
 
-    final newPage = _pageFactory.create(_currentConfiguration);
+    final newPage = _pageFactory.create(
+      _currentConfiguration.name,
+      _currentPageParameters,
+    );
 
     _updatePageParameters();
```
Comment (solid-vovabeloded) [inline metrics/web/lib/common/presentation/navigation/state/navigation_notifier.dart:223]: 🔨
It seems like we should update the page parameters firstly, and only then create a page using the updated page parameters. What do you think?

File: metrics/web/lib/common/presentation/navigation/metrics_page/metrics_page_factory.dart [hunk 4]
```diff
@@ -1,57 +1,69 @@
...
+    PageParametersModel pageParameters,
+  ) {
     switch (routeName) {
       case RouteName.loading:
         return MetricsPage(
           child: const LoadingPage(),
-          name: routePath,
+          routeName: routeName,
+          arguments: pageParameters,
+          name: routeName.value,
```
Comment (solid-vovabeloded) [inline metrics/web/lib/common/presentation/navigation/metrics_page/metrics_page_factory.dart:30]: :dry:
What do you think about refactoring this method to make it DRYer?

File: metrics/web/test/common/presentation/navigation/state/navigation_notifier_test.dart [hunk 5]
```diff
@@ -304,7 +306,8 @@ void main() {
         notifier.push(DefaultRoutes.projectGroups);
         notifier.pop();
 
-        verify(pageParametersFactory.create(currentConfiguration)).called(once);
+        verify(pageParametersFactory.create(currentConfiguration))
+            .called(equals(2));
```
Comment (solid-vovabeloded) [inline metrics/web/test/common/presentation/navigation/state/navigation_notifier_test.dart:310]: ✅ 
It seems like here we are expecting to get only one call of the `pageParametersFactory.create(currentConfiguration)`.

File: metrics/web/lib/common/presentation/navigation/state/navigation_notifier.dart [hunk 6]
```diff
@@ -215,7 +215,10 @@ class NavigationNotifier extends ChangeNotifier {
 
     _currentConfiguration = newConfiguration;
 
-    final newPage = _pageFactory.create(_currentConfiguration);
+    final newPage = _pageFactory.create(
+      _currentConfiguration.name,
+      _currentPageParameters,
+    );
```
Comment (solid-vovabeloded) [inline metrics/web/lib/common/presentation/navigation/state/navigation_notifier.dart:221]: ✅ 
Should we test that the `.push` and `.handleInitialRoutePath` methods use the given page factory to create a page? Also, I believe, we should test that these methods provide the correct parameters to the `MetricsPageFactory.create` method.

**Review by solid-vovabeloded — CHANGES_REQUESTED** [review]
Well done, @solid-maxim & @solid-tetianamovlian! I've added a couple of suggestions here - please take a look.

File: metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart [hunk 7]
```diff
@@ -2,167 +2,239 @@
...
     final routeName = RouteNameMock();
 
     tearDown(() {
-      reset(routeConfiguration);
       reset(routeName);
     });
 
     test(
-      ".create() returns the dashboard metrics page if the given route configuration is null",
+      ".create() returns the dashboard metrics page if the given route name is null",
```
Comment (solid-vovabeloded) [inline metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart:35]: 🏷️ 
```suggestion
      ".create() returns the metrics page with the dashboard page child if the given route name is null",
```
What do you think? Please, apply this suggestion to all similar situations.

File: metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart [hunk 8]
```diff
@@ -2,167 +2,239 @@
...
-        final page = metricsPageFactory.create(null);
+        final page = metricsPageFactory.create(null, pageParameters);
 
         expect(page.child, isA<DashboardPage>());
       },
     );
 
     test(
-      ".create() returns the loading metrics page with a name equals to the given route configuration path",
+      ".create() returns the dashboard metrics page with the name equals to the dashbord route name value if the given route name is null",
```
Comment (solid-vovabeloded) [inline metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart:44]: 🏷️ 
```suggestion
      ".create() returns the metrics page with the name equals to the dashboard route name value if the given route name is null",
```

File: metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart [hunk 9]
```diff
@@ -2,167 +2,239 @@
...
       () {
-        const routeConfiguration = DefaultRoutes.loading;
-        final expectedName = routeConfiguration.path;
+        final page = metricsPageFactory.create(null, pageParameters);
 
-        final actualName = metricsPageFactory.create(routeConfiguration).name;
+        expect(page.name, dashboardRouteName.value);
```
Comment (solid-vovabeloded) [inline metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart:48]: ✅ 
```suggestion
        expect(page.routeName, dashboardRouteName);
```
Does it make sense to test the `routeName` instead of `name`? What do you think? Please, apply this suggestion to all similar tests.

File: metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart [hunk 10]
```diff
@@ -2,167 +2,239 @@
...
+      },
+    );
 
-        expect(actualName, equals(expectedName));
+    test(
+      ".create() returns the dashboard metrics page with arguments equals to the given page parameters if the route name is null",
```
Comment (solid-vovabeloded) [inline metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart:53]: 🏷️ 
```suggestion
      ".create() returns the metrics page with arguments equals to the given page parameters if the given route name is null",
```
It seems like this method returns a `MetricsPage`, but not the `DashboardMetricsPage`. Please, apply this suggestion to all similar tests.

File: metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart [hunk 11]
```diff
@@ -2,167 +2,239 @@
...
+        final page = metricsPageFactory.create(routeName, null);
 
-        expect(actualName, equals(expectedName));
+        expect(page.child, isA<DashboardPage>());
       },
     );
 
     test(
-      ".create() returns the dashboard metrics page with a name equals to the given route configuration path",
+      ".create() returns the dashboard metrics page with a name equals to the given route name value if the route name is unknown",
```
Comment (solid-vovabeloded) [inline metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart:73]: 🏷️ 
```suggestion
      ".create() returns the metrics page with a name equals to the dashboard route name value if the given route name is unknown",
```

File: metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart [hunk 12]
```diff
@@ -2,167 +2,239 @@
...
+        final page = metricsPageFactory.create(dashboardRouteName, null);
 
-        expect(actualName, equals(expectedName));
+        expect(page.child, isA<DashboardPage>());
       },
     );
 
     test(
-      ".create() returns the dashboard metrics page with a name equals to the given route configuration path if the configuration name is unknown",
+      ".create() returns the dashboard metrics page with a name equals to the given route name value",
```
Comment (solid-vovabeloded) [inline metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart:104]: ✅ 
```suggestion
      ".create() returns the metrics page with a name equals to the given route name value",
```
Also, maybe we need only one such test? Meaning we can test that this method returns the metrics page with the right name on edge cases and then test that this method returns the metrics page with the given route name. What do you think? The same suggestion for `.create() returns the dashboard metrics page with arguments equals to the given page parameters` test.

File: metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_test.dart [hunk 13]
```diff
@@ -63,19 +85,24 @@ void main() {
     test(
       ".createRoute() returns a metrics page with the given route name",
```
Comment (solid-vovabeloded) [inline metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_test.dart:86]: 🏷️ 
```suggestion
      ".createRoute() returns a metrics page with the name equal to the given route name value",
```
Also, should we test the `name` getter of this class?

File: metrics/web/test/common/presentation/navigation/state/navigation_notifier_test.dart [hunk 14]
```diff
@@ -409,6 +412,24 @@ void main() {
       },
     );
 
+    test(
+      ".push() uses the given metrics page factory with correct parameters to create a metrics page",
```
Comment (solid-vovabeloded) [inline metrics/web/test/common/presentation/navigation/state/navigation_notifier_test.dart:416]: 🏷️ 
```suggestion
      ".push() uses the given metrics page factory to create a metrics page",
```
Please, apply this suggestion to all similar situations.

**Review by solid-vovabeloded — CHANGES_REQUESTED** [review]
Great job, @solid-maxim & @solid-tetianamovlian! I've added a few more of suggestions here - please take a look. Overall, the code looks great - thanks!

File: metrics/web/lib/common/presentation/navigation/metrics_page/metrics_page_factory.dart [hunk 15]
```diff
@@ -13,45 +15,68 @@ import 'package:metrics/project_groups/presentation/pages/project_group_page.dar
...
+
+    switch (routeConfigurationName) {
       case RouteName.loading:
-        return MetricsPage(
-          child: const LoadingPage(),
-          name: routePath,
-        );
+        child = const LoadingPage();
+        routeName = routeConfigurationName;
+        path = routeConfiguration.path;
```
Comment (solid-vladlaba) [inline metrics/web/lib/common/presentation/navigation/metrics_page/metrics_page_factory.dart:41]: :todo:
These two lines repeat in every case of the switch. It seems like we can improve the code here to reduce duplication or remove it at all. DM me if you cannot find the best way.

I suspect, we can select the `child` only and use the `default` case to nullify the incorrect values.

File: metrics/web/lib/common/presentation/navigation/metrics_page/metrics_page_factory.dart [hunk 16]
```diff
@@ -13,45 +15,68 @@ import 'package:metrics/project_groups/presentation/pages/project_group_page.dar
...
+        child = const LoginPage();
+        routeName = routeConfigurationName;
+        path = routeConfiguration.path;
+        break;
       case RouteName.dashboard:
-        return MetricsPage(
-          child: DashboardPage(),
-          name: routePath,
-        );
+        child = DashboardPage();
```
Comment (solid-vladlaba) [inline metrics/web/lib/common/presentation/navigation/metrics_page/metrics_page_factory.dart:49]: 🔨 
It seems like we can introduce the const constructor to both `DashboardPage` and `ProjectGroupPage`.

File: metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart [hunk 17]
```diff
@@ -16,153 +18,193 @@ import '../../../../test_utils/route_name_mock.dart';
...
 
     final metricsPageFactory = MetricsPageFactory();
-    final routeConfiguration = RouteConfigurationMock();
     final routeName = RouteNameMock();
-
-    tearDown(() {
-      reset(routeConfiguration);
-      reset(routeName);
-    });
+    final routeConfiguration = RouteConfigurationMock();
```
Comment (solid-vladlaba) [inline metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart:28]: ✅ 
It seems like we should reset mocks after every test case to avoid side effects and run test cases independently.

File: metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart [hunk 18]
```diff
@@ -16,153 +18,193 @@ import '../../../../test_utils/route_name_mock.dart';
...
-        final page = metricsPageFactory.create(null);
+        final page = metricsPageFactory.create(null, pageParameters);
 
         expect(page.child, isA<DashboardPage>());
       },
     );
 
     test(
-      ".create() returns the loading metrics page with a name equals to the given route configuration path",
+      ".create() returns the metrics page with the name equals to the route configuration dashboard path if the given route configuration is null",
```
Comment (solid-vladlaba) [inline metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart:42]: 🏷️ 
What do you think? To avoid excessive details.
```suggestion
      ".create() returns the metrics page with the name equals to the dashboard path if the given route configuration is null",
```

File: metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart [hunk 19]
```diff
@@ -16,153 +18,193 @@ import '../../../../test_utils/route_name_mock.dart';
...
-        final expectedName = routeConfiguration.path;
+        final page = metricsPageFactory.create(null, pageParameters);
+
+        expect(page.routeName, equals(dashboardRouteName));
+      },
+    );
 
-        final actualName = metricsPageFactory.create(routeConfiguration).name;
+    test(
+      ".create() returns the metrics page with the arguments equals to the given page parameters if the given route configuration is null",
```
Comment (solid-vladlaba) [inline metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart:60]: 🏷️ 
```suggestion
      ".create() returns the metrics page with the arguments equal to the given page parameters if the given route configuration is null",
```

File: metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart [hunk 20]
```diff
@@ -16,153 +18,193 @@ import '../../../../test_utils/route_name_mock.dart';
...
+        final page = metricsPageFactory.create(routeConfiguration, null);
 
-        expect(actualName, equals(expectedName));
+        expect(page.child, isA<DashboardPage>());
       },
     );
 
     test(
-      ".create() returns the project groups metrics page with a name equals to the given route configuration path",
+      ".create() returns the metrics page with a name equals to the route configuration dashboard path if the given route configuration name is unknown",
```
Comment (solid-vladlaba) [inline metrics/web/test/common/presentation/navigation/metrics_page/metrics_page_factory_test.dart:81]: 🏷️ 
Similar to the one from above.

File: metrics/web/test/common/presentation/navigation/state/navigation_notifier_test.dart [hunk 21]
```diff
@@ -940,6 +958,24 @@ void main() {
       },
     );
 
+    test(
+      ".handleInitialRoutePath() uses the given metrics page factory with correct parameters to create a metrics page",
```
Comment (solid-vladlaba) [inline metrics/web/test/common/presentation/navigation/state/navigation_notifier_test.dart:962]: 🏷️ 
This is a suggestion from the previous review by @solid-vovabeloded. Please recheck every suggestion carefully one more time. 
```suggestion
      ".handleInitialRoutePath() uses the given metrics page factory to create a metrics page",
```

**Review by solid-vladlaba — CHANGES_REQUESTED** [review]
Well done, @solid-maxim! I've added a couple of suggestions below - please let me know what you think!

**Review by solid-vladlaba — APPROVED** [review]
LGTM, thanks @solid-maxim @solid-tetianamovlian!:rocket:
